### PR TITLE
bpo-29654 : Support If-Modified-Since HTTP header (browser cache)

### PR DESCRIPTION
**Pull request by PierreQuentel**
The PR adds support of the If-Modified-Since HTTP request header. 

If the user agent sends this header and the url matches a file that was not modified after the value specified in the header, the server returns status code 304 (Not Modified).

A test for this feature (test_browser_cache) is added in test_httpservers.py.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username " requested by the form is the "Login name" field under "Your Details".
4. If you just signed the CLA, please **wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

File: Lib/http/server.py [hunk 1]
```diff
@@ -687,6 +688,21 @@ def send_head(self):
             self.send_error(HTTPStatus.NOT_FOUND, "File not found")
             return None
         try:
+            # Use browser cache if possible
+            if "If-Modified-Since" in self.headers:
```
Comment (vadmium) [inline Lib/http/server.py]: The RFC <https://tools.ietf.org/html/rfc7232#section-3.3> says you have to ignore this field if If-None-Match is also specified

File: Lib/http/server.py [hunk 2]
```diff
@@ -687,6 +688,21 @@ def send_head(self):
...
+                # compare If-Modified-Since and date of last file modification
+                fs = os.stat(path)
+                ims = email.utils.parsedate(self.headers["If-Modified-Since"])
```
Comment (vadmium) [inline Lib/http/server.py]: This email function accepts a timezone (e.g. from the documentation: -0500), and seems to ignore it. But it may make more sense for a HTTP server to either adjust the date according to the timezone (e.g. add 5 hours), or treat it as an error (required by RFC 7232).

File: Lib/test/test_httpservers.py [hunk 3]
```diff
@@ -444,6 +444,20 @@ def test_head(self):
         self.assertEqual(response.getheader('content-type'),
                          'application/octet-stream')
 
+    def test_browser_cache(self):
+        #constructs the path relative to the root directory of the HTTPServer
+        response = self.request(self.base_url + '/test')
+        self.check_status_and_reason(response, HTTPStatus.OK, data=self.data)
+        # get Last-Modified response heaer
```
Comment (vadmium) [inline Lib/test/test_httpservers.py]: Spelling: header. But this comment seems redundant; isn’t the code obvious enough?

File: Lib/test/test_httpservers.py [hunk 4]
```diff
@@ -444,6 +444,20 @@ def test_head(self):
...
+        last_modif = response.headers['Last-modified']
+        # send new request to the same url with request header 
+        # If-Modified-Since
+        from email.message import Message
+        headers = Message()
+        headers['If-Modified-Since'] = last_modif
+        response = self.request(self.base_url + '/test', headers=headers)
+        self.check_status_and_reason(response, HTTPStatus.NOT_MODIFIED)
```
Comment (vadmium) [inline Lib/test/test_httpservers.py]: Please repeat the test with an earlier date, verifying that a response body is returned.

File: Lib/http/server.py [hunk 5]
```diff
@@ -687,6 +688,21 @@ def send_head(self):
...
+                if ims is not None:
+                    ims_datetime = datetime.datetime(*ims[:7])
+                    ims_dtstring = ims_datetime.strftime("%d %b %Y %H:%M:%S")
+                    last_modif = datetime.datetime.utcfromtimestamp(
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: `ims_datetime` is naive datetime object, but `datetime.datetime.utcfromtimestamp()` creates aware datetime object.

File: Lib/http/server.py [hunk 6]
```diff
@@ -687,6 +688,21 @@ def send_head(self):
...
+                        fs.st_mtime).strftime("%d %b %Y %H:%M:%S")
+                    if last_modif <= ims_dtstring:
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: Why compare string representations rather than datetime objects or just tuples? The order of such strings is different from the order of datetime objects.

File: Lib/http/server.py [hunk 7]
```diff
@@ -1212,3 +1228,5 @@ def test(HandlerClass=BaseHTTPRequestHandler,
     else:
         handler_class = SimpleHTTPRequestHandler
     test(HandlerClass=handler_class, port=args.port, bind=args.bind)
+
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: Unrelated changes.

File: Lib/http/server.py [hunk 8]
```diff
@@ -687,6 +688,21 @@ def send_head(self):
             self.send_error(HTTPStatus.NOT_FOUND, "File not found")
             return None
         try:
+            # Use browser cache if possible
+            if "If-Modified-Since" in self.headers:
+                # compare If-Modified-Since and date of last file modification
+                fs = os.stat(path)
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: Use `os.fstat()`.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Please sign the PSF contributor agreement and add entries in Misc/NEWS and What's New file.

File: Lib/http/server.py (hunk 8 above)
Comment (vadmium) [reply Lib/http/server.py]: Maybe combine with the existing call below, too

File: Lib/http/server.py [hunk 9]
```diff
@@ -687,25 +687,30 @@ def send_head(self):
         except OSError:
             self.send_error(HTTPStatus.NOT_FOUND, "File not found")
             return None
+
+        fs = os.fstat(f.fileno())
+            
         try:
             # Use browser cache if possible
-            if "If-Modified-Since" in self.headers:
+            if "If-Modified-Since" in self.headers \
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: It is more preferable to use parenthesis rather than line continuation.

File: Lib/http/server.py [hunk 10]
```diff
@@ -687,25 +687,30 @@ def send_head(self):
...
+                    and not "If-None-Match" in self.headers:
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: `"If-None-Match" not in self.headers`

File: Lib/http/server.py [hunk 11]
```diff
@@ -687,25 +687,30 @@ def send_head(self):
...
-                fs = os.stat(path)
                 ims = email.utils.parsedate(self.headers["If-Modified-Since"])
                 if ims is not None:
-                    ims_datetime = datetime.datetime(*ims[:7])
-                    ims_dtstring = ims_datetime.strftime("%d %b %Y %H:%M:%S")
-                    last_modif = datetime.datetime.utcfromtimestamp(
-                        fs.st_mtime).strftime("%d %b %Y %H:%M:%S")
-                    if last_modif <= ims_dtstring:
+                    # If-Modified-Since is UTC, rounded to the second
+                    tzinfo = datetime.timezone(datetime.timedelta(hours=0))
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: `datetime.timezone.utc`

But maybe use naive datetime objects?

**Comment by PierreQuentel** [comment]
Thanks for all the feedback.

I have signed the CLA, it will probably be approved tomorrow. I will add the entries in Misc/NEWS and What's New.

In the second commit : 
- the datetime comparison now uses datetime objects in the UTC timezone, not their string representation. The time of file last modification must be rounded to the second because If-Modified-Since is also rounded
- if a `If-None-Match` header is present, `If-Modified-Since` is ignored - but at the moment, `If-None-Match` is not handled
- tests have been added

File: Lib/http/server.py [hunk 12]
```diff
@@ -687,25 +687,30 @@ def send_head(self):
...
+        fs = os.fstat(f.fileno())
+            
         try:
             # Use browser cache if possible
-            if "If-Modified-Since" in self.headers:
+            if "If-Modified-Since" in self.headers \
+                    and not "If-None-Match" in self.headers:
                 # compare If-Modified-Since and date of last file modification
-                fs = os.stat(path)
                 ims = email.utils.parsedate(self.headers["If-Modified-Since"])
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: It looks a problem to me that `email.utils.parsedate()` silently ignores timezone. Maybe use `email.utils.parsedate_to_datetime()` instead?

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Lib/test/test_httpservers.py [hunk 13]
```diff
@@ -444,6 +444,46 @@ def test_head(self):
...
+        headers['If-None-Match'] = "*"
+        response = self.request(self.base_url + '/test', headers=headers)
+        self.check_status_and_reason(response, HTTPStatus.OK)        
+
+        # with If-Modified-Since earlier than Last-Modified, must return 200
+        import datetime
+        import email.utils
+        dt = email.utils.parsedate(last_modif)
+        # build datetime object : one year before last modification
+        old_dt = [dt[0] - 1] + list(dt[1:7])
```
Comment (vadmium) [inline Lib/test/test_httpservers.py]: Assuming the date reflects when the test suite is run, this will fail on a leap day (e.g. 2020-02-29). Perhaps try something like “datetime(*old_dt) - timedelta(days=365)”.

File: Lib/http/server.py [hunk 14]
```diff
@@ -686,10 +687,30 @@ def send_head(self):
...
+                    and not "If-None-Match" in self.headers:
+                # compare If-Modified-Since and date of last file modification
+                ims = email.utils.parsedate(self.headers["If-Modified-Since"])
+                if ims is not None:
+                    # If-Modified-Since is UTC, rounded to the second
+                    tzinfo = datetime.timezone(datetime.timedelta(hours=0))
+                    ims_datetime = datetime.datetime(*ims[:7], tzinfo=tzinfo)
+                    # compare to UTC datetime of last modification, also 
+                    # rounded to the second
+                    mtime = int(fs.st_mtime)
```
Comment (vadmium) [inline Lib/http/server.py]: “Int” rounds towards 1970. The actual rounding that “date_time_string” uses seems to be a combination of rounding half to even microseconds (thanks to datetime), and then rounding down (always backwards in time) to seconds.

It might be more sensible to call datetime.replace(microseconds=0).

File: Lib/test/test_httpservers.py [hunk 15]
```diff
@@ -444,6 +444,46 @@ def test_head(self):
...
+        old_dt = datetime.datetime(*old_dt)
+        headers = Message()
+        headers['If-Modified-Since'] = email.utils.format_datetime(old_dt)
+        response = self.request(self.base_url + '/test', headers=headers)
+        self.check_status_and_reason(response, HTTPStatus.OK)
+
+        # build datetime object : one hours after last modification
+        new_dt = [dt[0]] + [dt[1] + 1] + list(dt[2:7])
```
Comment (vadmium) [inline Lib/test/test_httpservers.py]: Dt[1] is attribute tm_mon, but the comment suggests you want to adjust by one hour.

File: Lib/http/server.py (hunk 9 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: done

File: Lib/http/server.py (hunk 10 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: done too

File: Lib/http/server.py (hunk 12 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: Ok, the latest commit uses `parsedate_to_datetime()` and checks that it has a timezone set to UTC, which should be the case for the header (cf [RFC 7231](https://tools.ietf.org/html/rfc7231#section-7.1.1.1)). To compare with last modification time, both must be timezone-aware datetimes.

File: Lib/test/test_httpservers.py (hunk 13 above)
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py]: Correct, I have modified the test.

File: Lib/test/test_httpservers.py (hunk 15 above)
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py]: I changed the implementation with `timedelta(hours=1)` instead

File: Doc/whatsnew/3.7.rst [hunk 16]
```diff
@@ -110,6 +110,12 @@ urllib.parse
 adding `~` to the set of characters that is never quoted by default.
 (Contributed by Christian Theune and Ratnadeep Debnath in :issue:`16285`.)
 
+http.server
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: Move before `unittest.mock`. Modules should be lexicographically ordered.

File: Lib/http/server.py [hunk 17]
```diff
@@ -686,10 +687,30 @@ def send_head(self):
...
+                    # compare to UTC datetime of last modification, 
+                    # rounded to the second
+                    mtime = int(fs.st_mtime)
+                    last_modif = datetime.datetime.fromtimestamp(mtime, 
+                        ims.tzinfo)
+                    if last_modif <= ims:
+                        self.send_response(HTTPStatus.NOT_MODIFIED)
+                        self.end_headers()
+                        f.close()
+                        return
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: `return None`. If there are `return` statements with value, all `None` values should be returned explicitly.

File: Lib/http/server.py [hunk 18]
```diff
@@ -686,10 +687,30 @@ def send_head(self):
...
+        fs = os.fstat(f.fileno())
+            
         try:
+            # Use browser cache if possible
+            if ("If-Modified-Since" in self.headers
+                    and "If-None-Match" not in self.headers):
+                # compare If-Modified-Since and time of last file modification
+                ims = email.utils.parsedate_to_datetime(
+                    self.headers["If-Modified-Since"])
+                if (ims is not None and
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: This can be written in one line.

File: Lib/http/server.py [hunk 19]
```diff
@@ -686,10 +687,30 @@ def send_head(self):
         except OSError:
             self.send_error(HTTPStatus.NOT_FOUND, "File not found")
             return None
+
+        fs = os.fstat(f.fileno())
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: `f` is leaked if `os.fstat()` raises an exception.

File: Lib/http/server.py [hunk 20]
```diff
@@ -686,10 +687,30 @@ def send_head(self):
...
+            
         try:
+            # Use browser cache if possible
+            if ("If-Modified-Since" in self.headers
+                    and "If-None-Match" not in self.headers):
+                # compare If-Modified-Since and time of last file modification
+                ims = email.utils.parsedate_to_datetime(
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: `email.utils.parsedate_to_datetime()` can raise an exception. It might be better just to ignore the cache rather than return an error.

Needed a cache for ill-formed "If-Modified-Since" value.

File: Lib/http/server.py [hunk 21]
```diff
@@ -686,10 +687,30 @@ def send_head(self):
         except OSError:
             self.send_error(HTTPStatus.NOT_FOUND, "File not found")
             return None
+
+        fs = os.fstat(f.fileno())
+            
         try:
+            # Use browser cache if possible
+            if ("If-Modified-Since" in self.headers
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: HTTP headers are case-insensitive. Does this work with "if-modified-since"?

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Lib/http/server.py [hunk 22]
```diff
@@ -686,10 +687,30 @@ def send_head(self):
...
+                    and "If-None-Match" not in self.headers):
+                # compare If-Modified-Since and time of last file modification
+                ims = email.utils.parsedate_to_datetime(
+                    self.headers["If-Modified-Since"])
+                if (ims is not None and
+                        ims.tzinfo is datetime.timezone.utc):
```
Comment (vadmium) [inline Lib/http/server.py]: Maybe you should also allow tzinfo to be None, to support the obsolete “asctime” format from <https://tools.ietf.org/html/rfc7231#section-7.1.1.1> without a timezone (e.g. Sun Nov  6 08:49:37 1994)

File: Lib/http/server.py (hunk 21 above)
Comment (vadmium) [reply Lib/http/server.py]: Self.headers should be (similar to) an email.message.Message object, so I think it should work fine

Comment (PierreQuentel) [reply Lib/http/server.py]: I confirm that it's case-insensitive

File: Doc/whatsnew/3.7.rst (hunk 16 above)
Comment (PierreQuentel) [reply Doc/whatsnew/3.7.rst]: Done in commit 143e1e9

File: Lib/http/server.py (hunk 19 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: done

File: Lib/http/server.py (hunk 17 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: done

File: Lib/http/server.py (hunk 20 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: done in c3337f40207a402944b61e89b617b21599b0519e

File: Lib/http/server.py (hunk 18 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: done in c3337f40207a402944b61e89b617b21599b0519e

File: Lib/http/server.py (hunk 22 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: This should not happen often, but it is supported in commit c3337f40207a402944b61e89b617b21599b0519e

File: Lib/http/server.py [hunk 23]
```diff
@@ -686,12 +687,42 @@ def send_head(self):
...
         try:
+            fs = os.fstat(f.fileno())
+            # Use browser cache if possible
+            if ("If-Modified-Since" in self.headers
+                    and "If-None-Match" not in self.headers):
+                # compare If-Modified-Since and time of last file modification
+                try:
+                    ims = email.utils.parsedate_to_datetime(
+                        self.headers["If-Modified-Since"])
+                except:
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: Never use bare `except` unless you reraise the exception. This hides such exceptions as KeyboardInterrupt or MemoryError.

File: Lib/http/server.py [hunk 24]
```diff
@@ -686,12 +687,42 @@ def send_head(self):
...
+                    # ignore ill-formed values
+                    ims = None
+                if ims is not None:
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: Can `email.utils.parsedate_to_datetime()` return `None`? If it can't, just use `else`.

File: Lib/test/test_httpservers.py [hunk 25]
```diff
@@ -444,6 +444,45 @@ def test_head(self):
         self.assertEqual(response.getheader('content-type'),
                          'application/octet-stream')
 
+    def test_browser_cache(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_httpservers.py:457]: I think it would be better to split this test on few tests.

File: Lib/test/test_httpservers.py [hunk 26]
```diff
@@ -444,6 +444,45 @@ def test_head(self):
...
+        response = self.request(self.base_url + '/test', headers=headers)
+        self.check_status_and_reason(response, HTTPStatus.NOT_MODIFIED)
+
+        # if If-None-Match header is present, ignore If-Modified-Since
+        headers['If-None-Match'] = "*"
+        response = self.request(self.base_url + '/test', headers=headers)
+        self.check_status_and_reason(response, HTTPStatus.OK)        
+
+        # with If-Modified-Since earlier than Last-Modified, must return 200
+        import datetime
```
Comment (serhiy-storchaka) [inline Lib/test/test_httpservers.py]: Import at module level. Importing inside functions is used in exceptional cases.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Lib/test/test_httpservers.py (hunk 25 above)
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py:457]: Done in commit 42edfe3856e0ae683dbffe3f7a75b9daec0898e8.
The datetime of last modification is stored in method setUp and used in the browser cache tests. I also added a test for the Last-Modified header.

File: Lib/test/test_httpservers.py (hunk 26 above)
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py]: done in commit 42edfe3856e0ae683dbffe3f7a75b9daec0898e8

File: Lib/http/server.py (hunk 23 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: I added the `try` block because you said `parsedate_to_datetime()` could raise exceptions ; do you know which ones ? By looking at the code I see that it might raise `TypeError` because `_parsedate_tz` may return `None` and the line
```python
*dtuple, tz = _parsedate_tz(data)
```
would raise a `TypeError` in this case ; but is it the only one ?

Anyway, the code is itself inside a `try...except` block where the `except` block is just `except:`, so this is not making things much worse.

File: Lib/http/server.py (hunk 24 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: `email.utils.parsedate_to_datetime()` either returns a `datetime.datetime()` or raises an exception, in which case `ims` is set to `None`.

File: Lib/http/server.py (hunk 23 above)
Comment (vadmium) [reply Lib/http/server.py]: That TypeError looks like a bug in the implementation. Other exceptions:

```python-console
>>> parsedate_to_datetime("Sun, 99 Feb 2017 20:49:31")
ValueError: day is out of range for month
>>> parsedate_to_datetime("Sun, 26 Feb 2017 20:49:31 +12345678901234")
OverflowError: Python int too large to convert to C long
>>> parsedate_to_datetime("Mon, 20 Nov , 19:12:08 GMT")
IndexError: string index out of range
```

The point is the outer exception handler reraises the exception, but your handler swallows it.

Comment (PierreQuentel) [reply Lib/http/server.py]: Thanks, the exceptions have been specified in commit  1e68be9e1e68e5c483f01043ba0d081a2e538290

File: Doc/whatsnew/3.7.rst [hunk 27]
```diff
@@ -110,7 +117,6 @@ urllib.parse
 adding `~` to the set of characters that is never quoted by default.
 (Contributed by Christian Theune and Ratnadeep Debnath in :issue:`16285`.)
 
-
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: Unrelated change.

File: Lib/http/server.py [hunk 28]
```diff
@@ -686,12 +687,42 @@ def send_head(self):
...
+            if ("If-Modified-Since" in self.headers
+                    and "If-None-Match" not in self.headers):
+                # compare If-Modified-Since and time of last file modification
+                try:
+                    ims = email.utils.parsedate_to_datetime(
+                        self.headers["If-Modified-Since"])
+                except (TypeError, IndexError, OverflowError, ValueError):
+                    # ignore ill-formed values
+                    ims = None
+                if ims is not None:
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: Just `else`:
```
except (TypeError, IndexError, OverflowError, ValueError):
    # ignore ill-formed values
    pass
else:
```

File: Lib/http/server.py [hunk 29]
```diff
@@ -1212,3 +1243,4 @@ def test(HandlerClass=BaseHTTPRequestHandler,
     else:
         handler_class = SimpleHTTPRequestHandler
     test(HandlerClass=handler_class, port=args.port, bind=args.bind)
+
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: Don't add this line.

Comment (PierreQuentel) [reply Lib/http/server.py]: Done in commit fc20596

File: Lib/http/server.py (hunk 28 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: done in commit fc20596

File: Doc/whatsnew/3.7.rst (hunk 27 above)
Comment (PierreQuentel) [reply Doc/whatsnew/3.7.rst]: Sorry, I don't understand what I should do here

Comment (serhiy-storchaka) [reply Doc/whatsnew/3.7.rst]: You removed an empty line. Insert it back.

Comment (PierreQuentel) [reply Doc/whatsnew/3.7.rst]: Ok, done in commit 9436e45

**Comment by PierreQuentel** [comment]
Is there something left to do before the pull request can be merged ?

File: Doc/whatsnew/3.7.rst [hunk 30]
```diff
@@ -96,6 +96,13 @@ New Modules
 Improved Modules
 ================
 
+http.server
+-----------
+:class:`SimpleHTTPRequestHandler` supports the HTTP If-Modified-Since header.
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: `` :class:`~http.server.SimpleHTTPRequestHandler` ``

File: Misc/NEWS [hunk 31]
```diff
@@ -666,6 +666,9 @@ Library
 - Issue #24142: Reading a corrupt config file left configparser in an
   invalid state.  Original patch by Florian Höch.
 
+- bpo-29654 : Support If-Modified-Since HTTP header (browser cache). Patch
```
Comment (serhiy-storchaka) [inline Misc/NEWS]: Remove the space before colon. Use two spaces after sentence ending period. Move the entry to the start of the "Library" section (new entries are added at the start).

File: Lib/http/server.py [hunk 32]
```diff
@@ -100,6 +100,7 @@
 import socketserver
 import sys
 import time
+import datetime
```
Comment (serhiy-storchaka) [inline Lib/http/server.py]: Try to support alphabetical order of imports. I.e. add `import datetime` before `import email.utils`.

File: Lib/test/test_httpservers.py [hunk 33]
```diff
@@ -17,8 +17,11 @@
 import urllib.parse
 import html
 import http.client
+import email.message
```
Comment (serhiy-storchaka) [inline Lib/test/test_httpservers.py]: Try to support alphabetical order of imports.

File: Lib/http/server.py (hunk 32 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: Done in commit 708ff03. The order was not supported for argparse and copy...
I also changed to version to 0.7.

File: Lib/test/test_httpservers.py (hunk 33 above)
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py]: I don't understand. In the current version, the imports are not in alphabetic order : 
```python
from http.server import BaseHTTPRequestHandler, HTTPServer, \
     SimpleHTTPRequestHandler, CGIHTTPRequestHandler
from http import server, HTTPStatus

import os
import sys
import re
import base64
import ntpath
import shutil
import urllib.parse
import html
import http.client
import tempfile
import time
from io import BytesIO

import unittest
from test import support
```
There doesn't seem to be any logical order either - `http.client` is not grouped with `http.server` for instance.

Would you mind explaining more what you want and why ? It will be helpful for next Pull Requests.

Comment (serhiy-storchaka) [reply Lib/test/test_httpservers.py]: Just place `import email.utils` the first and `import datetime` after it. You can also move `import base64` before them.

Comment (serhiy-storchaka) [reply Lib/test/test_httpservers.py]: This is just a style nit, you shouldn't follow it if you have other reasons.

Comment (PierreQuentel) [reply Lib/test/test_httpservers.py]: I changed the order in commit 	db46671. I tried several orders but couldn't find something obvious. I prefer to leave Internet-related modules together, and `time` and `datetime` together.

... and thanks for the expression "style nit", I didn't know it :-)

File: Doc/whatsnew/3.7.rst (hunk 30 above)
Comment (PierreQuentel) [reply Doc/whatsnew/3.7.rst]: done in 6a58894

File: Misc/NEWS (hunk 31 above)
Comment (PierreQuentel) [reply Misc/NEWS]: done in 6a58894

**Comment by PierreQuentel** [comment]
Is there still anything to do ?

**Review by serhiy-storchaka — APPROVED** [review]
LGTM, but maybe @vadmium have other comments.

File: Lib/http/server.py [hunk 34]
```diff
@@ -80,13 +80,16 @@
 # (Actually, the latter is only true if you know the server configuration
 # at the time the request was made!)
 
-__version__ = "0.6"
+__version__ = "0.7"
```
Comment (vadmium) [inline Lib/http/server.py]: I don’t understand the `__version__` change. This is roughly documented as affecting SimpleHTTPRequestHandler.server_version, and the Server response header field. But the current 0.6 value has existed since 2001: revision 077153e. It seems safer to leave it alone unless we know what the numbers mean etc.

File: Doc/whatsnew/3.7.rst [hunk 35]
```diff
@@ -97,6 +97,13 @@ New Modules
 Improved Modules
 ================
 
+http.server
+-----------
+:class:`~http.server.SimpleHTTPRequestHandler` supports the HTTP
+If-Modified-Since header. The server returns the 304 response status if the
+target file was not modified after the datetime specified in the header.
```
Comment (vadmium) [inline Doc/whatsnew/3.7.rst]: Would it be better to just say “after the time specified in the header”? Datetime is just a Python implementation detail, not an English word or part of HTTP.

Also, I wonder if the details should go in the main documentation. Already <https://docs.python.org/dev/library/http.server.html#http.server.SimpleHTTPRequestHandler.do_GET> has quite a few details of the server behaviour.

**Review by vadmium — COMMENTED** [review]
The code changes look okay in general, and I support the idea of the change. I haven’t actually tested this out though (and am unlikely to do so for a while). Just left a couple minor thoughts.

File: Lib/http/server.py (hunk 34 above)
Comment (PierreQuentel) [reply Lib/http/server.py]: Done in 6dadf6e

File: Doc/whatsnew/3.7.rst (hunk 35 above)
Comment (PierreQuentel) [reply Doc/whatsnew/3.7.rst]: Replaced "datetime" by "time" in c389bf6

Comment (PierreQuentel) [reply Doc/whatsnew/3.7.rst]: Update of the `http.server` documentation proposed in 294e164

**Comment by serhiy-storchaka** [comment]
Do you still have comments @vadmium?

**Comment by vadmium** [comment]
Only other thing that I wonder about is if the documentation should say the Not-Modified handling was added in 3.7 (usually with “.. versionchanged::” markup). But if other people don’t think this needs to be added, I guess it doesn’t matter that much.

**Comment by PierreQuentel** [comment]
I think it's worth mentioning, for instance if a test suite expects a status 200 and gets 304 in version 3.7. It's not a very likely use case, but it may happen. I have added the "Changed in version 3.7" comment in c8b108f.

**Comment by PierreQuentel** [comment]
Can you merge the PR if there is nothing else to do ?

**Comment by serhiy-storchaka** [comment]
The Travis CI build failed: trailing whitespaces in `http.server.rst`.

**Comment by PierreQuentel** [comment]
Fixed in 16f9e11, all checks pass.

**Comment by PierreQuentel** [comment]
@serhiy-storchaka @vadmium 
Thank you for your patience, it was my first PR for this project and I missed things that are obvious to you.
